sql/driver/string_scanner.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package driver
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+)
+
+const (
+	dateFormat                    = "2006-01-02"
+	timestampWithOffsetZoneFormat = "2006-01-02 15:04:05.999999999-07:00"
+)
+
+var _ sql.Scanner = &StringScanner{}
+var _ fmt.Stringer = &StringScanner{}
+
+// StringScanner is a thin wrapper around sql.NullString that supports
+// scanning from time.Time.
+type StringScanner struct {
+	nullString sql.NullString
```

Comment:
> What does it mean for the string to be valid? In sql.NullString, Valid means "not-NULL".

That's exactly what I mean: `sql.NullString` makes it possible to distinguish null and non-null strings; this `driver.NullString` does not.

> it's also the point of disambiguation between DDate and DTimestamp.

Then the type name should reflect that. But even then, it's surprising that timestamps which happen to have all zeros for their sub-day fields will be printed with a different format. I think we need an additional wire type for this distinction, just like we do to distinguish durations from other numbers.